src/sentry/api/serializers/rest_framework/dashboard.py

```diff
@@ -0,0 +1,186 @@
+            raise serializers.ValidationError("One or more queries are required to create a widget")
+        return data
+
+
+class DashboardDetailsSerializer(serializers.Serializer):
+    # Is a string because output serializers also make it a string.
+    id = serializers.CharField(required=False)
+    title = serializers.CharField(required=False)
+    widgets = DashboardWidgetSerializer(many=True, required=False)
+
+    validate_id = validate_id
+
+    def update(self, instance, validated_data):
+        """
+        Update a dashboard, the connected widgets and queries
+
+        - Widgets in the dashboard currently, but not in widget_data will be removed.
+        - Widgets without ids will be created.
+        - Widgets with matching IDs will be updated.
+        - The order of the widgets will be updated based on the order in the request data.
+
+        Only call save() on this serializer from within a transaction or
+        bad things will happen
+        """
+        instance.title = validated_data.get("title", instance.title)
+        instance.save()
+
+        if "widgets" in validated_data:
+            self.update_widgets(instance, validated_data["widgets"])
+
+        return instance
+
+    def update_widgets(self, instance, widget_data):
+        widget_ids = [widget["id"] for widget in widget_data if "id" in widget]
+
+        existing_widgets = DashboardWidget.objects.filter(dashboard=instance, id__in=widget_ids)
+        existing_map = {widget.id: widget for widget in existing_widgets}
+
+        # Remove widgets that are not in the current request.
+        self.remove_missing_widgets(instance.id, widget_ids)
+
+        # Get new ordering start point.
+        next_order = get_next_dashboard_order(instance.id)
+
+        for i, data in enumerate(widget_data):
+            widget_id = data.get("id")
+            if widget_id and widget_id in existing_map:
+                # Update existing widget.
+                self.update_widget(existing_map[widget_id], data, next_order + i)
+            elif not widget_id:
+                # Create a new widget.
+                self.create_widget(instance, data, next_order + i)
+            else:
+                raise serializers.ValidationError(
+                    "You cannot use existing widgets that are not part of this dashboard."
+                )
+
+    def remove_missing_widgets(self, dashboard_id, keep_ids):
+        """
+        Removes current widgets belonging to dashboard not in keep_ids.
+        """
+        DashboardWidget.objects.filter(dashboard_id=dashboard_id).exclude(id__in=keep_ids).delete()
+
+    def create_widget(self, dashboard, widget_data, order):
+        widget = DashboardWidget.objects.create(
+            dashboard=dashboard,
+            display_type=widget_data["displayType"],
+            title=widget_data["title"],
+            order=order,
+        )
+        for i, query in enumerate(widget_data.pop("queries")):
```

Comment:
Can we use `bulk_create` here to avoid `N+1` queries?

Reply:
Good idea. I totally forgot about bulk_create.